rust/onnxruntime/src/session.rs

```diff
@@ -0,0 +1,747 @@
+    session_ptr: *mut sys::OrtSession,
+    allocator_ptr: *mut sys::OrtAllocator,
+    memory_info: MemoryInfo,
+    /// Information about the ONNX's inputs as stored in loaded file
+    pub inputs: Vec<Input>,
+    /// Information about the ONNX's outputs as stored in loaded file
+    pub outputs: Vec<Output>,
+}
+
+/// Information about an ONNX's input as stored in loaded file
+#[derive(Debug)]
+pub struct Input {
+    /// Name of the input layer
+    pub name: String,
+    /// Type of the input layer's elements
+    pub input_type: TensorElementDataType,
+    /// Shape of the input layer
+    ///
+    /// C API uses a i64 for the dimensions. We use an unsigned of the same range of the positive values.
+    pub dimensions: Vec<Option<u32>>,
+}
+
+/// Information about an ONNX's output as stored in loaded file
+#[derive(Debug)]
+pub struct Output {
+    /// Name of the output layer
+    pub name: String,
+    /// Type of the output layer's elements
+    pub output_type: TensorElementDataType,
+    /// Shape of the output layer
+    ///
+    /// C API uses a i64 for the dimensions. We use an unsigned of the same range of the positive values.
+    pub dimensions: Vec<Option<u32>>,
+}
+
+impl Input {
+    /// Return an iterator over the shape elements of the input layer
+    ///
+    /// Note: The member [`Input::dimensions`](struct.Input.html#structfield.dimensions)
+    /// stores `u32` (since ONNX uses `i64` but which cannot be negative) so the
+    /// iterator converts to `usize`.
+    pub fn dimensions(&self) -> impl Iterator<Item = Option<usize>> + '_ {
+        self.dimensions.iter().map(|d| d.map(|d2| d2 as usize))
+    }
+}
+
+impl Output {
+    /// Return an iterator over the shape elements of the output layer
+    ///
+    /// Note: The member [`Output::dimensions`](struct.Output.html#structfield.dimensions)
+    /// stores `u32` (since ONNX uses `i64` but which cannot be negative) so the
+    /// iterator converts to `usize`.
+    pub fn dimensions(&self) -> impl Iterator<Item = Option<usize>> + '_ {
+        self.dimensions.iter().map(|d| d.map(|d2| d2 as usize))
+    }
+}
+
+impl Drop for Session {
+    #[tracing::instrument]
+    fn drop(&mut self) {
+        debug!("Dropping the session.");
+        if self.session_ptr.is_null() {
+            error!("Session pointer is null, not dropping.");
+        } else {
+            unsafe { g_ort().ReleaseSession.unwrap()(self.session_ptr) };
+        }
+
+        self.session_ptr = std::ptr::null_mut();
+        self.allocator_ptr = std::ptr::null_mut();
+    }
+}
+
+unsafe impl Send for Session {}
+
+impl Session {
+    /// Run the input data through the ONNX graph, performing inference.
+    ///
+    /// Note that ONNX models can have multiple inputs; a `Vec<_>` is thus
+    /// used for the input data here.
+    pub fn run<'s, 't, 'm, TIn, TOut, D>(
+        &'s self,
+        input_arrays: Vec<Array<TIn, D>>,
+    ) -> Result<Vec<OrtOwnedTensor<'t, 'm, TOut, ndarray::IxDyn>>>
```

Comment:
This run API can not use multiple types of dimensions.
How do you resolve this problem?
I needed abstraction by trait in [VOICEVOX/onnxruntime-rs](https://github.com/VOICEVOX/onnxruntime-rs/blob/master/onnxruntime/src/session.rs#L508-L511) .

Reply:
So I am working on this. I have a branch here that does something similar to your trait. https://github.com/boydjohnson/onnxruntime/tree/try-update-bindings.

Reply:
That API's input_arrays is `Vec<impl ConstructTensor + Debug>` type.
However, I think that API seems make compile error like this codes.
```rust
let ar1 = ndarray::arr0(&(0 as i64));
let ar2 = ndarray::arr1(&[1]);
let v = vec![ar1, ar2]; // It will be mismatched types
session.run(v);
```
Therefore I think it needs implements from impl ConstructTensor to dyn ConstructTensor.

Reply:
Yeah, you are right. I am looking into this. It should be `&dyn ConstructTensor`.

Reply:
@qwerty2501 Take a look at what I have now for inputs and outputs. It should work for inputs of different types and outputs of different types.

Reply:
@boydjohnson I tried compile new sample code. But it has new compile error.
```rust
    let ar1 = ndarray::arr0(&(0 as i64));
    let ar2 = ndarray::arr1(&[1]);
    let v: Vec<&mut dyn ConstructTensor> = vec![&mut ar1, &mut ar2];
    session.run(v);
```
error message
```bash
error[E0277]: the trait bound `&i64: Utf8Data` is not satisfied
   --> onnxruntime/src/session.rs:763:49
    |
763 |     let v: Vec<&mut dyn ConstructTensor> = vec![&mut ar1, &mut ar2];
    |                                                 ^^^^^^^^ the trait `Utf8Data` is not implemented for `&i64`
    |
    = help: the following other types implement trait `Utf8Data`:
              &'a str
              String
note: required because of the requirements on the impl of `TypeToTensorElementDataType` for `&i64`
   --> onnxruntime/src/lib.rs:466:19
    |
466 | impl<T: Utf8Data> TypeToTensorElementDataType for T {
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^     ^
note: required because of the requirements on the impl of `ConstructTensor` for `ArrayBase<OwnedRepr<&i64>, Dim<[usize; 0]>>`
   --> onnxruntime/src/tensor/ort_input_tensor.rs:44:12
    |
44  | impl<T, D> ConstructTensor for Array<T, D>
    |            ^^^^^^^^^^^^^^^     ^^^^^^^^^^^
    = note: required for the cast to the object type `dyn ConstructTensor`
```

It seems unimplemented `TypeToTensorElementDataType` to &i64.
Should I implements `TypeToTensorElementDataType` to &i64 expressly?

Reply:
I think it is how `ar1` is constructed. Try `let mut ar1 = ndarray::arr0(0_i64)`.

Reply:
@qwerty2501 My latest push re-works the `Session::run` function. For your code sample it would now be
```
let ar1 = ndarray::arr0(0_i64);
let ar2 = ndarray::arr1(&[1]);
let v = vec![ar1.into(), ar2.into()];
session.run(v).unwrap();
```